crates/sparrow-scheduler/src/queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Provide the global and per-thread (local) work queues.
+//!
+//! These wrap the currently used crate to make it easy to swap in
+//! different implementations.
+
+pub trait Queue<T> {
+    /// Push an item to the *front* of the local queue.
+    ///
+    /// The item will be placed in the LIFO slot (next thing this thread runs).
+    /// If there was already an item in that slot, it will be displaced.
+    ///
+    /// If the local queue is full, it will move half of its items to the global queue.
+    ///
+    /// When called on a global queue, items will be pushed to that.
+    fn push(&mut self, item: T);
+
+    /// Push an item to the *end* of the local queue, skipping the LIFO slot.
+    ///
+    /// This can be used to give other tasks a chance to run. Otherwise, there’s
+    /// a risk that one task will completely take over a thread in a push-pop
+    /// cycle due to the LIFO slot.
+    ///
+    /// When called on a global queue, items will be pushed to that.
+    fn push_yield(&mut self, item: T);
```

Comment:
This trait was introduced so that we can run the logic of a pipeline from a worker (by passing the current worker queue) or from outside (by passing the global injector). This allows us to read from a Parquet file and push directly into the first transform, scheduling work with the global injector.